### ISSUE TITLE
[Feature]: An option to NOT shutdown worker after test failure

### ISSUE THREAD

**Issue by vitalets** [microsoft/playwright#32803]
### 🚀 Feature Request

Currently Playwright always [creates a new worker](https://playwright.dev/docs/test-parallel#worker-processes) if test fails:

> Workers are always shutdown after a [test failure](https://playwright.dev/docs/test-retries#failures) to guarantee pristine environment for following tests.

Although such behavior is quite reasonable, it has some downsides:
* performance impact: all worker fixtures re-run in new worker that can be expensive
* `beforeAll` also re-run that can be confusing for users, see https://github.com/microsoft/playwright/issues/22520
* retries of flaky tests can be not optimal, see https://github.com/microsoft/playwright/issues/23354

In most of my tests, I don't store any state in the worker. I just use test scoped fixtures that are re-created for every test anyway. For such cases, an option to re-use worker would significantly improve performance.

Strictly speaking, currently approach does not guarantee 100% pristine environment even for passed tests. In the following example `test 2` fails because `test 1` changes global state of regexp:
```js
const titleRe = /end-to-end testing/g;

test('test 1', async ({ page }) => {
  await page.goto('https://playwright.dev');
  const title = await page.title();
  expect(titleRe.test(title)).toEqual(true);
});

test('test 2', async ({ page }) => {
  await page.goto('https://playwright.dev');
  const title = await page.title();
  expect(titleRe.test(title)).toEqual(true);
});
```

### Example

I propose to add an option `recreateWorker` that can hold one of two values: 
 - `on-failure` (current behavior, default)
 - `never` (not shutdown worker on failure)

For example:
```ts
export default defineConfig({
  use: {
    recreateWorker: 'on-failure', // or 'never' 
  },
```

### Motivation

When tests don't store any state in worker, having an option to re-use worker after failures with greatly improve performance.

**Comment by matt-kxs** [comment]
Is there any workaround for this? I already put in work to get beforeAll / afterAll to execute once per file on first worker only, but was shocked to discover that the worker is destroyed on failure. This makes porting to playwright near impossible unless we disable parallelism.

**Comment by vitalets** [comment]
This is a workaround I've shared in another issue: https://github.com/microsoft/playwright/issues/22520#issuecomment-2391025061
It allows to run code once across workers. Suitable for `beforeAll` hooks, where we typically interested in the first invocation.
The approach does not work for `afterAll` logic, where we interested in the _last_ invocation. It requires more difficult trick.

**Comment by matt-kxs** [comment]
> > Is there any workaround for this?
> 
> This is a workaround I've shared in another issue: [#22520 (comment)](https://github.com/microsoft/playwright/issues/22520#issuecomment-2391025061) It allows to run code once across workers. Suitable for `beforeAll` hooks, where we typically interested in the first invocation. The approach does not work for `afterAll` logic, where we interested in the _last_ invocation. It requires more difficult trick.

Thanks for this! The .afterAll is the tricky part I can't get working well with the destroying worker on failure case. For now I've settled on the following:

1. Use my existing structure where I've forced .beforeAll / .afterAll to work like it does in most other frameworks where it only executes once per file
2. I have a command line runner that I've modified to run one file at a time and will run with multiple workers, retry=0, maxFailures = 1.
3. If the file has test failures, re-run but with workers=1, retry=2, maxFailures=100

So when things are good we get the full benefits of parallelism but when things fail, fall back to how playwright wants things to work (losing all the speed benefits unfortunately).

**Comment by ztesterparadise2** [comment]
Please, inform on the decision in respect of the issue. It really is a problem for test automation on our project

**Comment by alexsch01** [comment]
I made a PR to fix this issue, please let me know any feedback you may have

**Comment by ztesterparadise2** [comment]
> I made a PR to fix this issue, please let me know any feedback you may have

Tried your changes, the worker is indeed recreated - browser does not close on failure. Which is awesome news! Thanks!

**Comment by alexsch01** [comment]
> Tried your changes, the worker is indeed recreated - browser does not close on failure. Which is awesome news! Thanks!

Great to hear! Actually in your case that's the worker not being recreated

**Comment by pavelfeldman** [comment]
Note to third party contributors: please don't attempt to fix it, it is harder than it sounds and is very unlikely to happen. We will experiement with it and address if feasible.

**Comment by alexsch01** [comment]
> Note to third party contributors: please don't attempt to fix it, it is harder than it sounds and is very unlikely to happen. We will experiement with it and address if feasible.

this feature is extremely useful for a playwright project with config options `fullyParallel=false, retries=0, workers=1` and reusing the same `page` object throughtout all tests - such as https://github.com/alexsch01/TestProj_playwright

This would give Playwright the equivalent of `testIsolation: false` in Cypress

If at all possible, I would like to see that option in Playwright

**Comment by matt-kxs** [comment]
> Note to third party contributors: please don't attempt to fix it, it is harder than it sounds and is very unlikely to happen. We will experiement with it and address if feasible.

Does this mean the 1.52 tag is not meant to actually address the problem? As I'm sure you're aware, several people keep logging bugs for this not realizing its by design - an opt out functionality would be great for the several projects that don't have the ability to operate in fully isolated environments or where setup / teardown is expensive.

**Comment by bhawna0612** [comment]
Any update on this issue? Will this be fixed in the upcoming version 1.52? This is much needed feature. I am new to Playwright and found this behaviour and then stumbled to post that many people have requested this feature.

**Comment by cefn** [comment]
Let me share our motivating case for getting away from the worker shutdown feature. 

We aim to spawn a running webserver per test ___file___ and are defeated by this failure-handling behaviour. Consequence: the runner will often spawn a webserver per ___test___ 😱

As individual tests within a test file are executed sequentially we set up and tear down differently-mocked backends to create different upstream scenarios for the server before loading a page to test that code path. During the page load for that test, the requests are served by the test's own ephemeral backends. 

Running in playwright (in node) rather than in cypress (in the browser) means it's a few lines of code using the express API to rapidly bring up an ephemeral backend on a worker-specific port for the lifetime of just a single test, and kill it after the test. Server code paths can't be tested by intercepting the browser's network calls - the network calls we need to control are between the server and its backends - we are exercising the full server stack here. 

To support parallelism, port numbers for test-file-launched servers and their ephemeral backends are allocated on demand from a range according to the worker's parallel index env. This is a deconflicting and lazy-loading behaviour which has to be brokered per VM. When workers are force-terminated after failure this ends up being a new server launched for every failing test. There is no per-test-file hook that I know of in playwright. There are only per-test-file-per-worker hooks, so this design choice has a HUGE impact.

Leaning into the convention that individual test files are self-contained and run serially by default, our developers can optimise their test suites for parallelism by bringing up a server shared between tests in the same file. In typical test files we dynamically launch a server instance once during a `beforeAll` and tear down in `afterAll`.

Killing workers on failure makes this server-lazy-load logic re-run multiple times per file if there are failures, making it much slower to work through the test impact of some change. A typical example is some style change which causes e.g. the visual snapshot check to fail across all tests. Suddenly the test suite is not only failing, but is an order-of-magnitude slower to complete for reasons we can't explain or control, so long as we use the playwright runner.

A whole server for one test does ___function___ e.g. with fullyParallel:true (our servers take 5-10 seconds to start depending if dev build or production build). However, it is wasteful of resources and slow for test files which should re-use browser contexts and servers between tests.

These slowdowns feel like a large and unnecessary penalty for using the playwright runner. The restart-worker-on-failure is not present any other test runner that I know of. This maybe suggests it's not so aligned with test practice as the implementers assume. So migrating to the vitest runner (and using the Playwright API) is an option I guess. In our fully-isolated tests there is nothing that compromises a worker, browser or server that has experienced a failed test and I honestly don't understand what scenario we are defending from by being forced to restart servers on every failure, but I don't think it applies to us. Is there a link to some justification? Or is it just a by-product of historical implementation choices?

So in summary, server launches are expected to be shared within a whole serially-executed test file since server startup is relatively intensive compared to e.g. visiting a page. For performance we depend on multiple tests per file to all reuse a server within a single worker.  

We reached for the flag to turn off this behaviour but found we have no control of it, so we are faced with VERY slow test suites in some circumstances thanks to this feature.

**Comment by frasu** [comment]
> Does this mean the 1.52 tag is not meant to actually address the problem?

Not really because sadly this tag just disappeared.

**Comment by matt-kxs** [comment]
@pavelfeldman, @yury-s would it be possible to get an explanation as to why there's so much reluctance to support this feature? So far I've only seen comments like 'not likely' without any elaboration. Are there technical complications or is the playwright team just against it? Some clarity would help so we know whether to investigate using other runners, creating workarounds, etc.

**Comment by NikitaTihiy** [comment]
I’d like to vote for this feature because the current setup doesn’t work well for my project—or at least not as far as I know.

We’re testing a pretty complex Electron app, and its login process can’t be bypassed with storageState because it depends on a shared component that’s used across iOS, Android, and other platforms we develop our app for. Before running all the tests in a suite, we usually clear any data related to the app from all possible folders and then do a fresh login. We have multiple test suites, each using different accounts, with setups synced from our backend.

The issue is that if one of the tests fails, the beforeAll function runs again, meaning the login process happens all over after clearing. Right now, I’ve got a workaround—I add a tag like "@login" to the first test in a suite, and in the beforeAll function, I check if any tests include that tag so I can skip redundant steps. But if there’s a better way to handle this, I’m all ears.

TL;DR:
I’m looking for a way to clear all data just once per suite, without repeating the clearing and login if one of the tests fails (storageState is not an option of our app). If there’s already a way to do this without adding the feature, I’d love to hear about it.

**Comment by pankajebix** [comment]
> > I made a PR to fix this issue, please let me know any feedback you may have
> 
> Tried your changes, the worker is indeed recreated - browser does not close on failure. Which is awesome news! Thanks!

How you have fix this ? Please let me know

**Comment by pankajebix** [comment]
> > Note to third party contributors: please don't attempt to fix it, it is harder than it sounds and is very unlikely to happen. We will experiement with it and address if feasible.
> 
> this feature is extremely useful for a playwright project with config options `fullyParallel=false, retries=0, workers=1` and reusing the same `page` object throughtout all tests - such as https://github.com/alexsch01/TestProj_playwright
> 
> This would give Playwright the equivalent of `testIsolation: false` in Cypress
> 
> If at all possible, I would like to see that option in Playwright

getting error when i am using this.

**Comment by vitalets** [comment]
I’ve built a package that lets you cache data between workers.
If the approach proves to be effective, I think Playwright could keep re-creating workers, while the user decides which data to cache. More details here: https://github.com/microsoft/playwright/issues/22520#issuecomment-3167474277

**Comment by cefn** [comment]
TL;DR this arbitrary choice of suite behaviour because [playwright knows best](https://github.com/microsoft/playwright/blob/b08a8f395f7bd7aa63199615d359826a92b3fb25/docs/src/test-parallel-js.md#L22) causes suites with large numbers of failing tests to go from minutes (without launching a new full stack after every failing test) to hours (when every failure causes a completely unnecessary relaunch of the full stack beyond our control - new frontend server, new mocked backends, new browser instance).

There is no API for suite precursors other than `beforeAll`, which is therefore broken in playwright. A new server-side stack (as necessarily provisioned by `beforeAll` in our suites) simply isn't needed for each client-side test regardless if it fails or not, but sounds like we are shouting into the void and better people know best.

It's not uncommon for a small change to have a knock-on effect on multiple tests, and because of this uncompromising feature it can therefore take hours to get this feedback instead of minutes. This is not a small issue. In my view browser-automation and CI in general is about establishing efficient feedback about failures. For any case like ours the playwright runner has broken the core feature and doesn't want to fix it.

With no progress on being able to preserve VM and suite state and the integrity of `beforeAll` after any failed test, I am recommending future projects don't use the playwright runner which seems not fit for purpose. We will continue to use Playwright for browser automation which is great, but use `vitest` as the test runner given these unfixable flaws. In this way we can restore control of test programming to the test developers.
